components/experimental/src/duration/format.rs

```diff
@@ -0,0 +1,528 @@
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::{options::*, Duration, DurationFormatter, DurationSign};
+
+use super::validated_options::Unit;
+use core::fmt::Write;
+use fixed_decimal::{FixedDecimal, SignDisplay};
+use writeable::{PartsWrite, Writeable};
+
+pub mod parts {
+    use writeable::Part;
+
+    pub const HOUR: Part = Part {
+        category: "unit",
+        value: "hour",
+    };
+
+    pub const MINUTE: Part = Part {
+        category: "unit",
+        value: "minute",
+    };
+
+    pub const SECOND: Part = Part {
+        category: "unit",
+        value: "second",
+    };
+
+    pub const LITERAL: Part = Part {
+        category: "literal",
+        value: "literal",
+    };
+}
+
+/// The [`Writeable`] implementation that is returned by [`DurationFormatter::format`]. See
+/// the [`writeable`] crate for how to consume this.
+#[derive(Debug)]
+pub struct FormattedDuration<'l> {
+    pub(crate) fmt: &'l DurationFormatter,
+    pub(crate) duration: &'l Duration,
+}
+
+impl<'a> FormattedDuration<'a> {
+    /// Section 1.1.9
+    /// Formats numeric hours to sink. Requires hours formatting style to be either Numeric or TwoDigit.
+    fn format_numeric_hours<V: PartsWrite + ?Sized>(
+        &self,
+        hour_val: u64,
+        sign_displayed: &mut bool,
+        sink: &mut V,
+    ) -> core::fmt::Result {
+        // 1. Let hoursStyle be durationFormat.[[HoursStyle]].
+
+        // 2. Assert: hoursStyle is "numeric" or hoursStyle is "2-digit".
+        debug_assert!(
+            self.fmt.options.hour == FieldStyle::Numeric
+                || self.fmt.options.hour == FieldStyle::TwoDigit
+        );
+        // 3. Let result be a new empty List.
+        // 4. Let nfOpts be OrdinaryObjectCreate(null).
+        // 5. Let numberingSystem be durationFormat.[[NumberingSystem]].
+        // 6. Perform ! CreateDataPropertyOrThrow(nfOpts, "numberingSystem", numberingSystem).
+
+        let mut fd = FixedDecimal::from(hour_val);
+
+        // 7. If hoursStyle is "2-digit", then
+        if self.fmt.options.hour == FieldStyle::TwoDigit {
+            // a. Perform ! CreateDataPropertyOrThrow(nfOpts, "minimumIntegerDigits", 2𝔽).
+            fd.pad_start(2);
+        }
+
+        // 8. If signDisplayed is false, then
+        if *sign_displayed {
+            // a. Perform ! CreateDataPropertyOrThrow(nfOpts, "signDisplay", "never").
+            fd.apply_sign_display(SignDisplay::Never);
+        }
+
+        // 9. Let nf be ! Construct(%NumberFormat%, « durationFormat.[[Locale]], nfOpts »).
+        // 10. Let hoursParts be ! PartitionNumberPattern(nf, hoursValue).
+        // 11. For each Record { [[Type]], [[Value]] } part of hoursParts, do
+        // a. Append the Record { [[Type]]: part.[[Type]], [[Value]]: part.[[Value]], [[Unit]]: "hour" } to result.
+
+        let ffd = self.fmt.fdf.format(&fd);
+        sink.with_part(parts::HOUR, |e| ffd.write_to_parts(e))?;
```

Comment:
Nit: `e` usually means "error". Pick a different letter like `w`

```suggestion
        sink.with_part(parts::HOUR, |w| ffd.write_to_parts(w))?;
```